src/sage/categories/finite_fields.py

```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+
+            EXAMPLES::
+                sage: k = GF((3, 10))
+                sage: k.quadratic_non_residue().is_square()
+                False
+                sage: k = GF((2, 10))
+                sage: k in Fields()  # to let k be a finite field
+                True
+                sage: k.quadratic_non_residue()
+                Traceback (most recent call last):
+                ...
+                ValueError: There are no non-squares in finite fields of even order
+            """
+            # if the order is an even power of two
+            # then every element is a square
+            if self.order() % 2 == 0:
+                raise ValueError("There are no non-squares in finite fields of even order")
+            # uniformly randomly select elements for a non-square
+            # with probablity 1/2 for a non-square
+            element = self.random_element()
+            while element.is_square():
+                element = self.random_element()
+            return element
+
     class ElementMethods:
-        pass
+        def is_square(self):
+            """
+            Tests if the element is a square or has
+            a square root element
+
+            OUTPUT:
+            - `True` if the element is a square `False` if not
+
+            EXAMPLES::
+                sage: S.<x> = GF(5)[]
+                sage: f = S.irreducible_element(20)
+                sage: k.<y> = S.quotient_ring(f)
+                sage: k in Fields()
+                True
+                sage: k(2).is_square()
+                True
+                sage: k.quadratic_non_residue().is_square()
+                False
+            """
+            if self.is_zero():
+                return True
+            q = self.parent().order()
+            if q % 2  == 0:
+                return True
+            character = self**((q-1)//2)
+            is_square = character == self.parent().one()
+            return is_square
+
+        def _tonelli(self):
+            """
+            Returns a square root of the element if it exists
+            using tonelli's algorithm
+
+            OUTPUT:
+
+            - A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._tonelli()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._tonelli()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+            """
+            q = self.parent().cardinality()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            g = self.parent().quadratic_non_residue()
+            odd_order = (q - 1).odd_part()
+            even_exp = Integer.valuation(q-1, 2)
+            e = 0
+            for i in range(2, even_exp+1):
```

Comment:
I'm curious if using Sage's `IntegerRange` instead of the python `range` has any performance impact.